src/packages/auth/functions.ts

```diff
@@ -1,11 +1,75 @@
+import jwt from 'jsonwebtoken';
 import type { Response } from 'node-fetch';
 import fetch from 'node-fetch';
 import { getEnv } from '../../env.js';
-import type { LoginResponse, LogoutResponse, SessionResponse, UserResponse } from './types.js';
 import getLogger from '../../logger.js';
+import type {
+  JsonWebToken,
+  JwtPayload,
+  JwtSecret,
+  LoginResponse,
+  LogoutResponse,
+  SessionResponse,
+  UserResponse,
+} from './types.js';
 
 const logger = getLogger('packages/auth/functions');
 
+export function authorizationHeaderToToken(authorizationHeader: string | undefined | null): JsonWebToken | never {
+  if (authorizationHeader !== null && authorizationHeader !== undefined) {
+    if (authorizationHeader.startsWith('Bearer ')) {
+      const [, token] = authorizationHeader.split(' '); // Split out 'Bearer' prefix.
+      return token;
+    } else {
+      throw new Error(`Authorization header does not include 'Bearer' prefix`);
+    }
+  } else {
+    throw new Error(`Authorization header not found`);
+  }
+}
+
+export function decodeJwt(authorizationHeader: string | undefined): JwtPayload | null {
+  try {
+    const token = authorizationHeaderToToken(authorizationHeader);
+    const { HASURA_GRAPHQL_JWT_SECRET } = getEnv();
+    const { key }: JwtSecret = JSON.parse(HASURA_GRAPHQL_JWT_SECRET);
+    const options: jwt.VerifyOptions = { algorithms: ['HS256'] };
+    const jwtPayload = jwt.verify(token, key, options) as JwtPayload;
+    return jwtPayload;
+  } catch (e) {
+    console.error(e);
+    return null;
+  }
+}
+
+export function generateJwt(
+  username: string,
+  camToken: string,
+  defaultRole: string,
+  allowedRoles: string[],
+  otherClaims: Record<string, string | string[]> = {},
+): string | null {
+  try {
+    const { HASURA_GRAPHQL_JWT_SECRET } = getEnv();
+    const { key, type }: JwtSecret = JSON.parse(HASURA_GRAPHQL_JWT_SECRET);
+    const options: jwt.SignOptions = { algorithm: type as 'HS256' | 'RS512', expiresIn: '36h' };
+    const payload: JwtPayload = {
+      camToken,
+      'https://hasura.io/jwt/claims': {
+        'x-hasura-allowed-roles': allowedRoles,
+        'x-hasura-default-role': defaultRole,
```

Comment:
Can `x-hasura-user-id: username` be added? It'll help with [authorization](https://hasura.io/blog/hasura-authorization-system-through-examples/) and can be used as a default insert/update column value (ie: `last_updated_by`).

Reply:
Done